### PR TITLE
Improve the performance of --dyno-scope-bundled

### PR DESCRIPTION
**Pull request by mppf**
This PR improves the performance of the new scope resolver. We have observed performance problems with `--dyno-scope-bundled`. These are more pronounced with developer builds, but are also present with release builds. The magnitude of the performance problems presents a challenge to turning on `--dyno-scope-bundled` by default.

This table summarizes the score, with times in seconds:

|                                                                        | main | this PR |
|------------------------------------------------------------------------|------|---------|
|`chpl --dyno-scope-bundled hello.chpl` (quickstart with CHPL_DEVELOPER) | 16.1 | 12.8    |
|`chpl --dyno-scope-bundled hello.chpl` (default, no CHPL_DEVELOPER)     | 3.0  | 2.6     |
|`chpl hello.chpl` (quickstart with CHPL_DEVELOPER)                      | 9.7  | 9.5     |
|`chpl hello.chpl` (default, no CHPL_DEVELOPER)                          | 2.5  | 2.4     |

This PR takes the following steps in order to achieve this:
 * Changes BorrowedIdsWithName to copy IDs rather than referring to OwnedIdsWithName. The idea of referring to a list of overloads with the same name was always an optimization. However, it did not seem to help performance and it added some complexity. Along with this change, renamed BorrowedIdsWithName to MatchingIdsWithName. There is no longer a need for `std::vector<BorrowedIdsWithName>` -- `MatchingIdsWithName` covers that.
 * Adds a query to compute the public contents of a module. After the changes in PR #19306, the public symbols available in a module are always fixed and at a single level (for shadowing purposes). This is responsible for most of the speed improvement here for `--dyno-scope-bundled`.
 * Removes certain `public use`s in the standard library. The motivation here is to remove the possibility for modules that `public use` each other (potentially through an intermediate module). This case is difficult to support efficiently in the new query to compute the public contents of a module. Removing these `public use` statements can be viewed as a continuation of PR #16119.
 * Adds a Flags to IdAndFlags to track Module-ness and Type-ness. These are useful to optimize some of the code using scope lookup. In particular, such code can use these flags instead of using `parsing::idToTag`.
 * Adjusts the strategy for handling conditionals with variables (see PR #21140) and do-while statements (see PR #22564). Both of these have scoping behavior where the parent scope is not the scope for the parent ID. Before this PR, that was handled with `nextHigherScope` which used queries to check for these patterns. This PR changes the process of constructing a `Scope` to take these patterns into account, so that `scope->parentScope()` can be used even in these cases.

Future Work:
 * It might be possible to further improve the new scope resolver by caching the scopes that need to be traversed when looking for a symbol. I prototyped something that saves such a linearized view of scopes in https://github.com/mppf/chapel/tree/save-scopes-visited . However, it will take more work to get this functional, and it's currently unclear if this offers enough performance improvement to be worthwhile.

- [x] full comm=none testing
- [x] spot-checked that a program can compile and run with `CHPL_LOCALE_MODEL=gpu`

Reviewed by @DanilaFe - thanks!

### CODE REVIEW

File: frontend/lib/resolution/resolution-queries.cpp [hunk 1]
```diff
@@ -3006,38 +3006,35 @@ isInitialTypedSignatureApplicable(Context* context,
...
+  if (candidate.isType() && !ci.isParenless()) {
     // calling a type - i.e. type construction
     const Type* t = initialTypeForTypeDecl(context, candidateId);
     return ApplicabilityResult::success(typeConstructorInitial(context, t));
   }
 
-  // not a candidate
-  if (ci.isMethodCall() && isFormal(tag)) {
+  // formals aren't candidates
+  if (ci.isMethodCall() && !candidate.isMethodOrField()) {
```
Comment (DanilaFe) [inline frontend/lib/resolution/resolution-queries.cpp:3033]: I'm not sure how this is equivalent to the code before; how does this condition guarantee that the candidate is a formal? If not, should the comment be adjusted?

File: frontend/lib/resolution/scope-queries.cpp [hunk 2]
```diff
@@ -1175,6 +1187,13 @@ bool LookupHelper::doLookupInScope(const Scope* scope,
                                    llvm::ArrayRef<const Scope*> receiverScopes,
                                    UniqueString name,
                                    LookupConfig config) {
+
+  /*
+  if (name == "BaseLocale")
+    printf("Looking up BaseLocale in %s\n", scope->id().str().c_str());
+  if (name == "writef")
+    printf("Looking up writef in %s\n", scope->id().str().c_str());*/
+
```
Comment (DanilaFe) [inline frontend/lib/resolution/scope-queries.cpp:1196]: probably want to delete this

File: frontend/lib/resolution/scope-queries.cpp [hunk 3]
```diff
@@ -1254,52 +1275,91 @@ bool LookupHelper::doLookupInScope(const Scope* scope,
...
+  const ModulePublicSymbols* modPublicSyms = nullptr;
   const ResolvedVisibilityScope* r = nullptr;
-  if (checkUseImport) {
+
+  // Gather the ModulePublicSymbols if we are in a situation to use it.
+  // Note, private imports end up in shadow scope 0, but are private.
+  if (checkDecls && checkUseImport && isModule(scope->tag()) &&
+      skipPrivateVisibilities && allowCached) {
+    //printf("Looking up public symbols for module %s\n", scope->id().str().c_str());
```
Comment (DanilaFe) [inline frontend/lib/resolution/scope-queries.cpp:1291]: dead printf comment

File: frontend/lib/resolution/scope-queries.cpp [hunk 4]
```diff
@@ -1254,52 +1275,91 @@ bool LookupHelper::doLookupInScope(const Scope* scope,
...
     if (resolving && resolving->scope() == scope) {
       r = resolving;
     } else {
       r = resolveVisibilityStmts(context, scope);
     }
+
+    // don't check use/import if there is no ResolvedVisibilityScope
+    // since that would imply there are no use/imports
+    if (r == nullptr && !scope->autoUsesModules())
+      checkUseImport = false;
```
Comment (DanilaFe) [inline frontend/lib/resolution/scope-queries.cpp:1307]: Why is this needed if `doLookupInImportsAndUses` already skips for null `r`?

File: frontend/lib/resolution/scope-queries.cpp [hunk 5]
```diff
@@ -1692,7 +1759,8 @@ static bool lookupInScopeViz(Context* context,
                             context, scope, /* receiverScopes */ {},
                             resolving, name, config,
                             checkedScopes, result,
-                            idForWarnings);
+                            idForWarnings,
+                            /* allowCached */ false);
```
Comment (DanilaFe) [inline frontend/lib/resolution/scope-queries.cpp:1763]: Why is `allowCached` false here?

File: frontend/lib/resolution/scope-queries.cpp [hunk 6]
```diff
@@ -1880,18 +1951,31 @@ getSymbolsAvailableInScopeQuery(Context* context,
...
     auto visStmtSymbols =
       getSymbolsAvailableInScopeQuery(context, vis.scope(), &vis);
     for (auto& pair : visStmtSymbols) {
-      toReturn.try_emplace(pair.first, pair.second);
+      UniqueString name = pair.first;
+      const MatchingIdsWithName& ids = pair.second;
+      auto end = ids.end();
+      for (auto cur = ids.begin(); cur != end; ++cur) {
+        toReturn[name].appendIfMatching(cur.curIdAndFlags(),
+                                              flags, excludeNothing);
```
Comment (DanilaFe) [inline frontend/lib/resolution/scope-queries.cpp:1964]: nit, but wrongly indented

File: frontend/lib/resolution/scope-queries.cpp [hunk 7]
```diff
@@ -3031,9 +3113,11 @@ GatherMentionedModules::lookupAndGather(const Scope* scope,
   // if we aren't in a method, use a faster strategy to find the module
   if (!inMethod) {
     const Scope* resultScope = nullptr;
-    for (const auto& bids : v) {
-      for (const auto& id : bids) {
-        if (resultScope == nullptr && parsing::idIsModule(context, id)) {
+    auto end = v.end();
+    for (auto cur = v.begin(); cur != end; ++cur) {
```
Comment (DanilaFe) [inline frontend/lib/resolution/scope-queries.cpp:3117]: I see; this is what creates the need for a custom iterator. I think this can be cleaned up using an "adapter", but probably not in this PR. I'd rather have two ways to iterate using the "nice" `for` loop.

File: frontend/lib/resolution/scope-queries.cpp [hunk 8]
```diff
@@ -3463,7 +3543,7 @@ emitMultipleDefinedSymbolErrorsQuery(Context* context, const Scope* scope) {
                         LOOKUP_IMPORT_AND_USE |
                         LOOKUP_SKIP_SHADOW_SCOPES;
   for (auto name : namesDefinedMultiply) {
-    std::vector<BorrowedIdsWithName> v =
+    MatchingIdsWithName v =
```
Comment (DanilaFe) [inline frontend/lib/resolution/scope-queries.cpp:3546]: Consider `auto` in cases like these

File: frontend/lib/resolution/scope-queries.cpp [hunk 9]
```diff
@@ -3520,6 +3600,108 @@ void emitMultipleDefinedSymbolErrors(Context* context, const Scope* scope) {
...
+                        LOOKUP_SKIP_SHADOW_SCOPES;
+  for (auto name : namesDefined) {
+    CheckedScopes lookupCheckedScopes;
+    MatchingIdsWithName lookupResult;
+
+    /*
+    if (scope->id().str() == "LocaleModelHelpFlat" &&
+        name == "writef") {
+      gdbShouldBreakHere();
+    }*/
```
Comment (DanilaFe) [inline frontend/lib/resolution/scope-queries.cpp:3631]: TBD (to be deleted)

File: frontend/lib/resolution/scope-queries.cpp [hunk 10]
```diff
@@ -3520,6 +3600,108 @@ void emitMultipleDefinedSymbolErrors(Context* context, const Scope* scope) {
...
+   */
+static const owned<ModulePublicSymbols>&
+publicSymbolsForModuleQuery(Context* context, const Scope* modScope) {
+  QUERY_BEGIN(publicSymbolsForModuleQuery, context, modScope);
+
+  //if (modScope->id().str() == "MemConsistency")
+  //  gdbShouldBreakHere();
+
+  //printf("publicSymbolsForModuleQuery %s begins\n",
+  //       modScope->id().str().c_str());
```
Comment (DanilaFe) [inline frontend/lib/resolution/scope-queries.cpp:3669]: TBD here too, and a bit below

File: frontend/lib/resolution/scope-types.cpp [hunk 11]
```diff
@@ -19,14 +19,58 @@
...
+                       bool isModule, bool isType)
+  : id_(std::move(id)) {
+
+  // setup the flags
+  Flags flags = 0;
+  if (isPublic) {
+    flags |= PUBLIC;
+  } else {
+    flags |= NOT_PUBLIC;
+  }
```
Comment (DanilaFe) [inline frontend/lib/resolution/scope-types.cpp:45]: What if you added a macro, like:

```C++
  #define BOOL_TO_FLAG(BOOL, NAME) \
    if (BOOL) {\
      flags |= NAME;\
    } else {\
      flags |= NOT_##NAME;\
    }
  BOOL_TO_FLAG(isPublic, PUBLIC);
  BOOL_TO_FLAG(isMethodOrField, METHOD_FIELD);
  BOOL_TO_FLAG(isParenfulFunction, PARENFUL_FUNCTION);
  BOOL_TO_FLAG(isMethod, METHOD);
  BOOL_TO_FLAG(isModule, MODULE);
  BOOL_TO_FLAG(isType, TYPE);
```

File: frontend/lib/resolution/scope-types.cpp [hunk 12]
```diff
@@ -179,72 +238,129 @@ OwnedIdsWithName::borrow(IdAndFlags::Flags filterFlags,
...
-    return BorrowedIdsWithName(*this, idv, filterFlags, excludeFlagSet);
+    // Found a visible ID! Return a MatchingIds referring to the whole thing
+    return MatchingIdsWithName(*this, filterFlags, excludeFlagSet);
   }
 
-  // No ID was visible, so we can't borrow.
-  return chpl::empty;
+  // No ID was visible, so return empty.
+  return MatchingIdsWithName();
 }
```
Comment (DanilaFe) [inline frontend/lib/resolution/scope-types.cpp:255]: Commented out; delete?

File: frontend/test/resolution/testScopeResolve.cpp [hunk 13]
```diff
@@ -1332,6 +1332,8 @@ static void test33() {
 
 
 int main() {
+  test6();
```
Comment (DanilaFe) [inline frontend/test/resolution/testScopeResolve.cpp:1335]: Why does this re-run `test6`? Does it still need to?

File: modules/internal/ChapelLocale.chpl [hunk 14]
```diff
@@ -27,8 +27,15 @@
 module ChapelLocale {
 
   public use LocaleModel as _; // let LocaleModel refer to the class
+  /*public import LocaleModel.{LocaleModel,
+                             RootLocale,
+                             localeModelHasSublocales,
+                             localeModelPartitionsIterationOnSublocales};*/
```
Comment (DanilaFe) [inline modules/internal/ChapelLocale.chpl:33]: TBD?

**Review by DanilaFe — APPROVED** [review]
To be clear, is the "difficult to support efficiently" case still supported? In case user modules create cycles?

**Comment by mppf** [comment]
> To be clear, is the "difficult to support efficiently" case still supported? In case user modules create cycles?

AFAIK, yes. It is just slower. This is handled by `publicSymbolsForModuleIfNotRunning` which checks if the query is running for the module in question.

File: frontend/include/chpl/resolution/scope-types.h [hunk 15]
```diff
@@ -355,62 +368,42 @@ class BorrowedIdsWithName {
...
+  MatchingIdsWithName(IdAndFlags idv) {
+    idvs_.push_back(std::move(idv));
   }
+
  public:
   /**
     Iterator that skips invisible entries from the list of borrowed IDs.
    */
-  class BorrowedIdsWithNameIter {
+  class MatchingIdsWithNameIter {
```
[the remark this answers is not in the document]
Comment (mppf) [reply frontend/include/chpl/resolution/scope-types.h:390]: The iterator defaults to yielding IDs rather than IdAndFlags

File: frontend/include/chpl/resolution/scope-types.h [hunk 16]
```diff
@@ -566,6 +553,12 @@ class BorrowedIdsWithName {
  */
 using DeclMap = std::unordered_map<UniqueString, OwnedIdsWithName>;
 
+bool lookupInDeclMap(const DeclMap& declared,
+                     UniqueString name,
+                     MatchingIdsWithName& result,
+                     IdAndFlags::Flags filterFlags,
+                     const IdAndFlags::FlagSet& excludeFlags);
```
[the remark this answers is not in the document]
Comment (mppf) [reply frontend/include/chpl/resolution/scope-types.h:560]: We have `using DeclMap = std::unordered_map<UniqueString, OwnedIdsWithName>;` and I can't add methods to `unordered_map`. It might be worthwhile to make `DeclMap` a class, but I don't think that's called for yet.

File: frontend/lib/resolution/resolution-queries.cpp (hunk 1 above)
Comment (mppf) [reply frontend/lib/resolution/resolution-queries.cpp:3033]: Formals aren't methods or fields. I'll adjust the comment in any case.

File: frontend/lib/resolution/scope-queries.cpp (hunk 5 above)
Comment (mppf) [reply frontend/lib/resolution/scope-queries.cpp:1763]: Having the caching turned off here simplifies things by keeping the potential for query recursion for mutually recursive modules within a single query here. `publicSymbolsForModule` uses the queries that build `ResolvedVisibilityScope`, and the `false` here prevents the queries that build `ResolvedVisibilityScope` from using `publicSymbolsForModule`.

I'll add a comment about this.

File: frontend/lib/resolution/scope-queries.cpp (hunk 8 above)
Comment (mppf) [reply frontend/lib/resolution/scope-queries.cpp:3546]: It's nice to put the type so less needs to be consulted to see what the code is doing.

File: frontend/lib/resolution/scope-queries.cpp (hunk 7 above)
Comment (mppf) [reply frontend/lib/resolution/scope-queries.cpp:3117]: In some ways, it'd be more natural for `for (const auto& x : myMatchingIdsWithName)` to yield `IdAndFlags`. However needing `ID` is common. I will add a TODO comment about this to `MatchingIdsWithName`.

**Comment by bradcray** [comment]
Nice!
